packages_rs/nextclade/src/tree/tree_find_nearest_node.rs

```diff
@@ -3,35 +3,67 @@ use crate::analyze::letter_ranges::NucRange;
 use crate::analyze::nuc_sub::NucSub;
 use crate::tree::tree::{AuspiceTree, AuspiceTreeNode};
 use crate::utils::range::Range;
+use itertools::Itertools;
 
+#[derive(Clone, Debug)]
 pub struct TreeFindNearestNodeOutput<'node> {
   pub node: &'node AuspiceTreeNode,
   pub distance: i64,
+  pub confidence: f64,
 }
 
 /// For a given query sample, finds nearest node on the reference tree (according to the distance metric)
-pub fn tree_find_nearest_node<'node>(
+pub fn tree_find_nearest_nodes<'node>(
   tree: &'node AuspiceTree,
   qry_nuc_subs: &[NucSub],
   qry_missing: &[NucRange],
   aln_range: &Range,
-) -> TreeFindNearestNodeOutput<'node> {
+) -> Vec<TreeFindNearestNodeOutput<'node>> {
   // Iterate over tree nodes and calculate distance metric between the sample and each node
-  tree
+  let nodes_by_distance = tree
     .iter_depth_first_preorder()
     .map(|(_, node)| {
       let distance = tree_calculate_node_distance(node, qry_nuc_subs, qry_missing, aln_range);
-      TreeFindNearestNodeOutput { node, distance }
+      TreeFindNearestNodeOutput {
+        node,
+        distance,
+        confidence: 0.0,
+      }
     })
+    .sorted_by_key(|out| out.distance)
+    .collect_vec();
+
+  match nodes_by_distance.first() {
+    // The first node in the list sorted by distance is one of the nearest nodes (nodes that all have minimum distance)
+    Some(first_nearest_node) => {
+      // Find all nearest nodes (nodes with the same distance as the first node)
+      let min_distance = first_nearest_node.distance;
+      let nearest_node_candidates = nodes_by_distance
+        .into_iter()
+        .filter(|node| node.distance == min_distance)
+        .collect_vec();
 
-    // Find nearest node (having minimum distance)
-    .min_by_key(|out| out.distance)
+      // Sort nearest nodes further
+      tree_sort_nearest_nodes(&nearest_node_candidates)
+    }
 
-    // If none can be found, return default (e.g. if tree is empty, which is unlikely)
-    .unwrap_or(TreeFindNearestNodeOutput {
+    // Unlikely case: if there's no nodes, return parent
+    None => vec![TreeFindNearestNodeOutput {
       node: &tree.tree,
       distance: 0,
-    })
+      confidence: 1.0,
+    }],
+  }
+}
+
+/// Reorder equivalently near nodes according to a metric, from the best to the worst.
+/// The first node in the resulting list will be considered "the most nearest" among equally nearest nodes
+/// and will be used for clade assignment.
+fn tree_sort_nearest_nodes<'node>(
+  nearest_node_candidates: &[TreeFindNearestNodeOutput<'node>],
+) -> Vec<TreeFindNearestNodeOutput<'node>> {
+  // TODO: actually reorder the list or create a new one, and calculate placement confidence for each node
+  nearest_node_candidates.to_vec()
 }
```

Comment:
This function needs to be implemented

Here we receive all candidate nearest node, (all having equal distance, which is the minimum distance, and we need to order the list according to our needs.